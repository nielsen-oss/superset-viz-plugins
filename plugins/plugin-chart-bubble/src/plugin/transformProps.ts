/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps, DataRecord } from '@superset-ui/core';
import { Metric, SortingType } from '@superset-viz-plugins/plugin-chart-waterfall/src/plugin/utils';
import { LegendPosition } from '@superset-viz-plugins/plugin-chart-waterfall/src/types';

export type QueryData = {
  [key: string]: number | string;
};

type FormData = {
  xAxisColumn: string;
  xAxisLabel: string;
  yAxisLabel: string;
  yAxisLabelAngle: string;
  yAxisLogScale: boolean;
  xAxisTickLabelAngle: string;
  xAxisLogScale: boolean;
  numbersFormat: string;
  colorScheme: string;
  legendPosition: LegendPosition;
  maxNumberOfLegends: string;
  showGridLines: boolean;
  bubbleSize: boolean;
  entity: string;
  series: string;
  xAxis: { label: string } | string;
  yAxis: { label: string } | string;
  zAxis: { label: string } | string;
};

export default function transformProps(chartProps: ChartProps) {
  const { width, height, formData, queriesData } = chartProps;
  const data = queriesData[0].data as QueryData[];

  const {
    xAxisColumn,
    numbersFormat,
    legendPosition,
    maxNumberOfLegends,
    showGridLines,
    bubbleSize,
    xAxisLogScale,
    yAxisLogScale,
    colorScheme,
    entity,
    series,
  } = formData as FormData;

  return {
    xAxisDataKey: xAxisColumn,
    xAxisLabel: formData.xAxisLabel ?? '',
    yAxisLabel: formData.yAxisLabel ?? '',
    xAxisTickLabelAngle: -Number(formData.xAxisTickLabelAngle ?? 45),
    yAxisLabelAngle: -Number(formData.yAxisLabelAngle),
    xAxisLogScale,
    yAxisLogScale,
    width,
    height,
    bubbleSize: bubbleSize ?? 1000,
    legendPosition,
    maxNumberOfLegends,
    showGridLines,
    numbersFormat,
    colorScheme,
    data,
    entity,
    series,
    xAxis: formData.xAxis?.label ?? formData.xAxis,
    yAxis: formData.yAxis?.label ?? formData.yAxis,
    zAxis: formData.zAxis?.label ?? formData.zAxis,
  };
}
