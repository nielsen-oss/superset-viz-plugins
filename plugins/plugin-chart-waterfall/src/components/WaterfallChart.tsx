/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { FC, useCallback, useEffect, useState } from 'react';
import { getNumberFormatter, styled, t } from '@superset-ui/core';
import {
  Bar,
  BarChart,
  CartesianGrid,
  ContentRenderer,
  LabelList,
  LabelProps,
  Legend,
  LegendProps,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import WaterfallTick from './WaterfallTick';
import { getChartStyles, LEGEND, LegendPosition, renderLabel } from './utils';
import WaterfallBar from './WaterfallBar';
import WaterfallTooltip from './WaterfallTooltip';

type WaterfallStylesProps = {
  height: number;
  width: number;
};

export type BarValue = [number, number];

export type WaterfallChartData = {
  [key: string]: string | boolean | number | BarValue;
};

export type WaterfallChartProps = {
  xAxisDataKey?: string;
  dataKey: string;
  legendPosition: LegendPosition;
  error?: string;
  numbersFormat?: string;
  height: number;
  resetFilters?: Function;
  onBarClick?: Function;
  width: number;
  data?: WaterfallChartData[];
};

const Styles = styled.div<WaterfallStylesProps>`
  height: ${({ height }) => height}px;
  width: ${({ width }) => width}px;
`;

const Error = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${({ theme }) => theme.gridUnit * 4}px;
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  color: ${({ theme }) => theme.colors.warning.dark1};
  background-color: ${({ theme }) => theme.colors.warning.light1};
`;

const Notification = styled.div`
  cursor: pointer;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${({ theme }) => theme.gridUnit * 4}px;
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  color: ${({ theme }) => theme.colors.info.dark1};
  background-color: ${({ theme }) => theme.colors.info.light1};
`;

const WaterfallChart: FC<WaterfallChartProps> = props => {
  const {
    onBarClick = () => {},
    xAxisDataKey,
    dataKey,
    numbersFormat,
    data,
    height,
    width,
    error,
    legendPosition,
  } = props;
  const [notification, setNotification] = useState<string | null>(null);
  const [updater, setUpdater] = useState<number>(0);

  const forceUpdate = useCallback(() => setUpdater(Math.random()), []);
  useEffect(() => {
    forceUpdate();
  }, [forceUpdate, props]);

  const handleBarClick = (barData: WaterfallChartData) => {
    if (window.location.pathname.includes('/explore')) {
      onBarClick(barData);
      setNotification(t('Bar was clicked, filter will be emitted on a dashboard'));
    }
  };
  const closeNotification = () => setNotification(null);
  const formatter = getNumberFormatter(numbersFormat);
  const { chartMargin, legendStyle } = getChartStyles(legendPosition);

  return (
    <Styles height={height} width={width}>
      {notification && <Notification onClick={closeNotification}>{notification}</Notification>}
      {error ? (
        <Error>{error}</Error>
      ) : (
        <div>
          <BarChart width={width} height={height} margin={chartMargin} data={data} barCategoryGap={0} key={updater}>
            <Legend
              wrapperStyle={legendStyle}
              verticalAlign={legendPosition as LegendProps['verticalAlign']}
              iconType="circle"
              iconSize={10}
              payload={LEGEND}
            />
            <CartesianGrid vertical={false} />
            <XAxis dataKey={xAxisDataKey} dy={10} angle={-45} tick={WaterfallTick} interval={0} />
            <YAxis tickFormatter={formatter} />
            <Tooltip content={<WaterfallTooltip formatter={formatter} />} />
            <Bar
              dataKey={dataKey}
              shape={props => <WaterfallBar {...props} numberOfBars={data?.length} />}
              onClick={handleBarClick}
            >
              <LabelList
                dataKey={dataKey}
                position="top"
                content={(renderLabel(formatter) as unknown) as ContentRenderer<LabelProps>}
              />
            </Bar>
          </BarChart>
        </div>
      )}
    </Styles>
  );
};

export default WaterfallChart;
