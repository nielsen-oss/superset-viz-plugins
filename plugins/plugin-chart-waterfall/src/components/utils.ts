/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { supersetTheme, t } from '@superset-ui/core';
import { LegendPayload } from 'recharts';
import { BarValue } from './WaterfallChart';

export const MIN_LABEL_MARGIN = 20;
export const MIN_SYMBOL_WIDTH_FOR_TICK_LABEL = 7;

export const LEGEND: LegendPayload[] = [
  { value: t('Increase'), color: supersetTheme.colors.success.base, id: 'increase', type: 'square' },
  { value: t('Decrease'), color: supersetTheme.colors.error.base, id: 'decrease', type: 'square' },
  { value: t('Total'), color: supersetTheme.colors.info.base, id: 'total', type: 'square' },
];

export enum LegendPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
}

export const renderLabel = (formatter: Function) => ({ value }: { value: BarValue }) =>
  `${formatter(value?.[1] - value?.[0])}`;

export const BOTTOM_PADDING = 25;

export const getChartStyles = (legendPosition: LegendPosition, yAxisWidth: number) => {
  let legendStyle: object = {
    paddingBottom: 20,
  };
  let chartMargin: object = { bottom: BOTTOM_PADDING, left: yAxisWidth + 10 };
  if (legendPosition === LegendPosition.BOTTOM) {
    legendStyle = {
      paddingTop: BOTTOM_PADDING,
    };
    chartMargin = { left: yAxisWidth + 10, top: 20 };
  }
  return {
    legendStyle,
    chartMargin,
  };
};
