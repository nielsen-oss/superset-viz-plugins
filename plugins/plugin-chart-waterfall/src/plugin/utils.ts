/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { t } from '@superset-ui/core';
import { WaterfallChartData } from '../components/WaterfallChart';
import { QueryData } from './transformProps';

export enum SortingType {
  ASC = 'ASC',
  DESC = 'DESC',
}

const groupDataByPeriod = (data: QueryData[], periodColumn: string) =>
  data.reduce((acc, cur) => {
    const period = cur[periodColumn] as string;
    const periodData = acc.get(period) || [];
    periodData.push(cur);
    acc.set(period, periodData);
    return acc;
  }, new Map<string, QueryData[]>());

const addTotalValueToPeriod = (
  valuesOfPeriod: QueryData[],
  xAxisColumn: string,
  periodColumn: string,
  valueColumn: string,
  key: string,
) => {
  // Calc total per period
  const sum = valuesOfPeriod.reduce((acc, cur) => acc + (cur[valueColumn] as number), 0);
  // Push total per period to the end of period values array
  valuesOfPeriod.push({
    [xAxisColumn]: key,
    [periodColumn]: '__TOTAL__',
    [valueColumn]: sum,
  });
};

const removeRedundantValuesInPeriod = (newVal: QueryData[], periodCounter: number) =>
  periodCounter === 0 ? [newVal[newVal.length - 1]] : newVal;

const findDiffOfPeriodsValue = (
  valueOfPeriod: QueryData,
  valueColumn: string,
  groupedData: Map<string, QueryData[]>,
  keys: string[],
  periodCounter: number,
  xAxisColumn: string,
): QueryData => ({
  ...valueOfPeriod,
  [valueColumn]:
    (valueOfPeriod[valueColumn] as number) -
    ((groupedData
      ?.get(keys[periodCounter - 1])
      ?.find(prevPeriodValue => prevPeriodValue[xAxisColumn] === valueOfPeriod[xAxisColumn])?.[
      valueColumn
    ] as number) ?? 0),
});

export const convertDataForRecharts = (
  periodColumn: string,
  xAxisColumn: string,
  valueColumn: string,
  data: QueryData[],
  orderByChange: SortingType,
  useOrderByChange: boolean,
) => {
  // Group by period (temporary map)
  const groupedData = groupDataByPeriod(data, periodColumn);

  let resultData: QueryData[] = [];
  let periodCounter = 0;
  groupedData.forEach((valuesOfPeriod, periodKey) => {
    const keys = [...groupedData.keys()];
    let newValuesOfPeriod: QueryData[] =
      periodCounter === 0
        ? valuesOfPeriod
        : valuesOfPeriod.map(valueOfPeriod =>
            findDiffOfPeriodsValue(valueOfPeriod, valueColumn, groupedData, keys, periodCounter, xAxisColumn),
          );

    addTotalValueToPeriod(newValuesOfPeriod, xAxisColumn, periodColumn, valueColumn, periodKey);
    newValuesOfPeriod = removeRedundantValuesInPeriod(newValuesOfPeriod, periodCounter);
    // Add client sort by change
    if (useOrderByChange) {
      newValuesOfPeriod.sort((a, b) =>
        a[periodColumn] === '__TOTAL__'
          ? 1
          : (orderByChange === SortingType.ASC ? 1 : -1) * ((a[valueColumn] as number) - (b[valueColumn] as number)),
      );
    }

    resultData = resultData.concat(newValuesOfPeriod);
    periodCounter += 1;
  });
  return resultData;
};

export const createReChartsBarValues = (
  rechartsData: QueryData[],
  valueColumn: keyof QueryData,
  periodColumn: keyof QueryData,
): WaterfallChartData[] => {
  // Create ReCharts values array of deltas for bars
  const resultData = rechartsData.map((cur: QueryData, index: number) => {
    let totalSumUpToCur = 0;
    for (let i = 0; i < index; i++) {
      // Ignore calculation on period column
      if (rechartsData[i][periodColumn] !== '__TOTAL__' || i === 0) {
        totalSumUpToCur += rechartsData[i][valueColumn] as number;
      }
    }

    const lastPeriod = totalSumUpToCur;
    const thisPeriod = totalSumUpToCur + (cur[valueColumn] as number);
    const change = thisPeriod - lastPeriod;
    const changePercentage = (change / lastPeriod) * 100;
    if (cur[periodColumn] === '__TOTAL__') {
      return {
        ...cur,
        thisPeriod,
        __TOTAL__: true,
        [valueColumn]: [0, totalSumUpToCur || cur[valueColumn]],
      } as WaterfallChartData;
    }
    return {
      ...cur,
      lastPeriod,
      thisPeriod,
      change,
      changePercentage,
      [valueColumn]: [lastPeriod, thisPeriod],
    } as WaterfallChartData;
  });

  return resultData;
};

export const MAX_FORM_CONTROLS = 50;

export const SortingTypeNames = {
  [SortingType.ASC]: t('Ascending'),
  [SortingType.DESC]: t('Descending'),
};
