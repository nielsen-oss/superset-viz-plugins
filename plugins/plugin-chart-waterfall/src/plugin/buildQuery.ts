/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { buildQueryContext, QueryFormData } from '@superset-ui/core';
import { BinaryOperator, SetOperator } from '@superset-ui/core/lib/query/types/Operator';
import { SortingType } from './utils';

// Not correctly imported form node_modules, so add it here
export type QueryFormExtraFilter = {
  col: string;
} & (
  | {
      op: BinaryOperator;
      val: string;
    }
  | {
      op: SetOperator;
      val: string[];
    }
);

export default function buildQuery(formData: QueryFormData) {
  return buildQueryContext(formData, baseQueryObject => {
    const orderby: [string, boolean][] = [];
    if (formData.use_order_by_period_column) {
      orderby.push([formData?.period_column, formData.order_by_period_column === SortingType.ASC]);
    }
    if (formData.use_order_by_x_axis_column) {
      orderby.push([formData?.x_axis_column, formData.order_by_x_axis_column === SortingType.ASC]);
    }
    if (formData.use_order_by_metric) {
      orderby.push([formData?.metric?.label, formData.order_by_metric === SortingType.ASC]);
    }

    return [
      {
        ...baseQueryObject,
        orderby,
        groupby: [formData.x_axis_column, formData.period_column],
      },
    ];
  });
}
