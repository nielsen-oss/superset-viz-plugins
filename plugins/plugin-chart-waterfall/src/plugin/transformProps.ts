/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps } from '@superset-ui/core';
import { WaterfallChartData } from '../components/WaterfallChart';
import { convertDataForRecharts, createReChartsBarValues, SortingType } from './utils';
import { LegendPosition } from '../components/utils';

type Metric = {
  label: string;
};

export type QueryData = {
  [key: string]: number | string;
};

type FormData = {
  xAxisColumn: string;
  periodColumn: string;
  queryFields: { metric: string };
  metric: Metric;
  numbersFormat: string;
  legendPosition: LegendPosition;
  orderByChange: SortingType;
  useOrderByChange: boolean;
};

export default function transformProps(
  chartProps: ChartProps,
): {
  dataKey: string;
  onBarClick: () => null;
  numbersFormat: string;
  data: WaterfallChartData[];
  xAxisDataKey: string;
  width: number;
  resetFilters: () => null;
  legendPosition: LegendPosition;
  height: number;
} {
  const { width, height, formData, queriesData } = chartProps;

  const {
    periodColumn,
    xAxisColumn,
    metric,
    numbersFormat,
    legendPosition,
    orderByChange,
    useOrderByChange,
  } = formData as FormData;

  const valueColumn = metric.label;
  const data = queriesData?.[0]?.data as QueryData[];

  const rechartsData = convertDataForRecharts(
    periodColumn,
    xAxisColumn,
    valueColumn,
    data,
    orderByChange,
    useOrderByChange,
  );

  const resultData = createReChartsBarValues(rechartsData, valueColumn, periodColumn);

  return {
    dataKey: valueColumn,
    xAxisDataKey: xAxisColumn,
    width,
    height,
    legendPosition,
    numbersFormat,
    data: resultData,
    onBarClick: () => null,
    resetFilters: () => null,
  };
}
