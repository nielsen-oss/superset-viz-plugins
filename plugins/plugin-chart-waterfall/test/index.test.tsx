/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import * as recharts from 'recharts';
import { supersetTheme, ThemeProvider } from '@superset-ui/core';
import { processNumbers } from '../src/plugin/utils';
import WaterfallChartPlugin from '../src';
import transformProps from '../src/plugin/transformProps';
import { legendTop } from './__mocks__/waterfallProps';
import WaterfallChart, { WaterfallChartData } from '../src/components/WaterfallChart';

jest.mock('recharts');

describe('plugin-chart-waterfall', () => {
  const BarChart = jest.fn(props => <div {...props} />);
  const CartesianGrid = jest.fn(() => <div />);
  const Legend = jest.fn(() => <div />);
  const Tooltip = jest.fn(() => <div />);
  const XAxis = jest.fn(() => <div />);
  const YAxis = jest.fn(() => <div />);
  const Bar = jest.fn(() => <div />);
  beforeEach(() => {
    // Recharts still have some UNSAFE react functions that failing test
    jest.spyOn(console, 'warn').mockImplementation(() => null);

    jest.clearAllMocks();

    // @ts-ignore
    recharts.CartesianGrid = CartesianGrid;
    // @ts-ignore
    recharts.BarChart = BarChart;
    // @ts-ignore
    recharts.Legend = Legend;
    // @ts-ignore
    recharts.Tooltip = Tooltip;
    // @ts-ignore
    recharts.XAxis = XAxis;
    // @ts-ignore
    recharts.YAxis = YAxis;
    // @ts-ignore
    recharts.Bar = Bar;
  });

  const getWrapper = (props: object) =>
    render(
      <ThemeProvider theme={supersetTheme}>
        {/*
       // @ts-ignore (no need emulate all props) */}
        <WaterfallChart {...transformProps(props)} />
      </ThemeProvider>,
    );

  it('exists', () => {
    expect(WaterfallChartPlugin).toBeDefined();
  });

  it('Chart with legend top', () => {
    getWrapper(legendTop);
    expect({
      BarChartProps: BarChart.mock.calls[0],
      CartesianGridProps: CartesianGrid.mock.calls[0],
      LegendProps: Legend.mock.calls[0],
      TooltipProps: Tooltip.mock.calls[0],
      XAxisProps: XAxis.mock.calls[0],
      YAxisProps: YAxis.mock.calls[0],
      BarProps: Bar.mock.calls[0],
    }).toMatchSnapshot();
  });

  describe('processNumbers', () => {
    const mockData = ([
      { metric: [-0.1114, 0] },
      { metric: [-0.1115, 0] },
      { metric: [-0.1116, 0] },
      { metric: [-0.1, 0] },
      { metric: [0, 0] },
      { metric: [0.1, 0] },
      { metric: [0.1114, 0] },
      { metric: [0.1115, 0] },
      { metric: [0.1116, 0] },
      { metric: [0.123, 0] },
      { metric: [0.123456, 0] },
      { metric: [123, 0] },
      { metric: [123456, 0] },
      { metric: [123456789, 0] },
      { metric: [1234567891011121314151617181920, 0] },
    ] as unknown) as WaterfallChartData[];
    it('non-adaptive', () => {
      expect(
        processNumbers(([{ metric: [123, 456] }] as unknown) as WaterfallChartData[], 'metric', 'SOME_FORMAT', '3'),
      ).toEqual([{ metric: [123, 456] }]);
    });
    it('adaptive with digits for different numbers', () => {
      expect(processNumbers(mockData, 'metric', 'SMART_NUMBER', '3')).toMatchSnapshot();
    });
    it('adaptive with digits without precision', () => {
      expect(processNumbers(mockData, 'metric', 'SMART_NUMBER')).toMatchSnapshot();
    });
  });
});
