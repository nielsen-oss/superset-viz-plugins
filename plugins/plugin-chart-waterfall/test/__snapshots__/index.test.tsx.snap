// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-chart-waterfall Chart with legend top 1`] = `
Object {
  "BarChartProps": Array [
    Object {
      "barCategoryGap": 0,
      "children": Array [
        <mockConstructor
          iconSize={10}
          iconType="square"
          payload={
            Array [
              Object {
                "color": "#5AC189",
                "id": "increase",
                "type": "square",
                "value": "Increase",
              },
              Object {
                "color": "#E04355",
                "id": "decrease",
                "type": "square",
                "value": "Decrease",
              },
              Object {
                "color": "#66BCFE",
                "id": "total",
                "type": "square",
                "value": "Total",
              },
            ]
          }
          verticalAlign="top"
          wrapperStyle={
            Object {
              "paddingBottom": 20,
            }
          }
        />,
        <mockConstructor
          horizontal={true}
          vertical={false}
        />,
        <mockConstructor
          angle={-45}
          dataKey="period_type"
          dy={10}
          height={6}
          interval={0}
          tick={[Function]}
        />,
        <mockConstructor
          domain={
            Array [
              29,
              134,
            ]
          }
          tickFormatter={[Function]}
          type="number"
        />,
        <mockConstructor
          content={
            <WaterfallTooltip
              formatter={[Function]}
            />
          }
        />,
        <mockConstructor
          dataKey="COUNT(volume_for_roi)"
          onClick={[Function]}
          shape={[Function]}
        >
          <LabelList
            content={[Function]}
            dataKey="COUNT(volume_for_roi)"
            position="top"
            valueAccessor={[MockFunction]}
          />
        </mockConstructor>,
      ],
      "data": Array [
        Object {
          "COUNT(volume_for_roi)": Array [
            29,
            122,
          ],
          "__TOTAL__": true,
          "change": 0,
          "period": "__TOTAL__",
          "period_type": "Total Period",
          "thisPeriod": 122,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            122,
            105,
          ],
          "change": -17,
          "changePercentage": -13.934426229508196,
          "lastPeriod": 122,
          "lastPeriodValue": 61,
          "period": "2018",
          "period_type": "Custom",
          "thisPeriod": 105,
          "thisPeriodValue": 44,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            105,
            86,
          ],
          "change": -19,
          "changePercentage": -18.095238095238095,
          "lastPeriod": 105,
          "lastPeriodValue": 61,
          "period": "2018",
          "period_type": "Standard",
          "thisPeriod": 86,
          "thisPeriodValue": 42,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            29,
            86,
          ],
          "__TOTAL__": true,
          "change": -36,
          "period": "__TOTAL__",
          "period_type": "2018",
          "thisPeriod": 50,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            86,
            86,
          ],
          "change": 0,
          "changePercentage": 0,
          "lastPeriod": 86,
          "lastPeriodValue": 44,
          "period": "2017",
          "period_type": "Custom",
          "thisPeriod": 86,
          "thisPeriodValue": 44,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            86,
            86,
          ],
          "change": 0,
          "changePercentage": 0,
          "lastPeriod": 86,
          "lastPeriodValue": 42,
          "period": "2017",
          "period_type": "Standard",
          "thisPeriod": 86,
          "thisPeriodValue": 42,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            29,
            86,
          ],
          "__TOTAL__": true,
          "change": 0,
          "period": "__TOTAL__",
          "period_type": "2017",
          "thisPeriod": 86,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            86,
            63,
          ],
          "change": -23,
          "changePercentage": -26.744186046511626,
          "lastPeriod": 86,
          "lastPeriodValue": 44,
          "period": "2020",
          "period_type": "Custom",
          "thisPeriod": 63,
          "thisPeriodValue": 21,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            63,
            40,
          ],
          "change": -23,
          "changePercentage": -36.507936507936506,
          "lastPeriod": 63,
          "lastPeriodValue": 42,
          "period": "2020",
          "period_type": "Standard",
          "thisPeriod": 40,
          "thisPeriodValue": 19,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            29,
            40,
          ],
          "__TOTAL__": true,
          "change": -46,
          "period": "__TOTAL__",
          "period_type": "2020",
          "thisPeriod": -6,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            40,
            39,
          ],
          "change": -1,
          "changePercentage": -2.5,
          "lastPeriod": 40,
          "lastPeriodValue": 21,
          "period": "2019",
          "period_type": "Custom",
          "thisPeriod": 39,
          "thisPeriodValue": 20,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            39,
            39,
          ],
          "change": 0,
          "changePercentage": 0,
          "lastPeriod": 39,
          "lastPeriodValue": 19,
          "period": "2019",
          "period_type": "Standard",
          "thisPeriod": 39,
          "thisPeriodValue": 19,
        },
        Object {
          "COUNT(volume_for_roi)": Array [
            29,
            39,
          ],
          "__TOTAL__": true,
          "change": -1,
          "period": "__TOTAL__",
          "period_type": "2019",
          "thisPeriod": 38,
        },
      ],
      "height": 400,
      "margin": Object {
        "bottom": 25,
        "left": 11,
      },
      "width": 800,
    },
    Object {},
  ],
  "BarProps": Array [
    Object {
      "children": <LabelList
        content={[Function]}
        dataKey="COUNT(volume_for_roi)"
        position="top"
        valueAccessor={[MockFunction]}
      />,
      "dataKey": "COUNT(volume_for_roi)",
      "onClick": [Function],
      "shape": [Function],
    },
    Object {},
  ],
  "CartesianGridProps": Array [
    Object {
      "horizontal": true,
      "vertical": false,
    },
    Object {},
  ],
  "LegendProps": Array [
    Object {
      "iconSize": 10,
      "iconType": "square",
      "payload": Array [
        Object {
          "color": "#5AC189",
          "id": "increase",
          "type": "square",
          "value": "Increase",
        },
        Object {
          "color": "#E04355",
          "id": "decrease",
          "type": "square",
          "value": "Decrease",
        },
        Object {
          "color": "#66BCFE",
          "id": "total",
          "type": "square",
          "value": "Total",
        },
      ],
      "verticalAlign": "top",
      "wrapperStyle": Object {
        "paddingBottom": 20,
      },
    },
    Object {},
  ],
  "TooltipProps": Array [
    Object {
      "content": <WaterfallTooltip
        formatter={[Function]}
      />,
    },
    Object {},
  ],
  "XAxisProps": Array [
    Object {
      "angle": -45,
      "dataKey": "period_type",
      "dy": 10,
      "height": 6,
      "interval": 0,
      "tick": [Function],
    },
    Object {},
  ],
  "YAxisProps": Array [
    Object {
      "domain": Array [
        29,
        134,
      ],
      "tickFormatter": [Function],
      "type": "number",
    },
    Object {},
  ],
}
`;

exports[`plugin-chart-waterfall processNumbers adaptive with digits for different numbers 1`] = `
Array [
  Object {
    "metric": Array [
      -0.111,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.112,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.112,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.1,
      0,
    ],
  },
  Object {
    "metric": Array [
      0,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.1,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.111,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.112,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.112,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.123,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.123,
      0,
    ],
  },
  Object {
    "metric": Array [
      123,
      0,
    ],
  },
  Object {
    "metric": Array [
      123456,
      0,
    ],
  },
  Object {
    "metric": Array [
      123456789,
      0,
    ],
  },
  Object {
    "metric": Array [
      1.2345678910111213e+30,
      0,
    ],
  },
]
`;

exports[`plugin-chart-waterfall processNumbers adaptive with digits without precision 1`] = `
Array [
  Object {
    "metric": Array [
      -0.1114,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.1115,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.1116,
      0,
    ],
  },
  Object {
    "metric": Array [
      -0.1,
      0,
    ],
  },
  Object {
    "metric": Array [
      0,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.1,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.1114,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.1115,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.1116,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.123,
      0,
    ],
  },
  Object {
    "metric": Array [
      0.123456,
      0,
    ],
  },
  Object {
    "metric": Array [
      123,
      0,
    ],
  },
  Object {
    "metric": Array [
      123456,
      0,
    ],
  },
  Object {
    "metric": Array [
      123456789,
      0,
    ],
  },
  Object {
    "metric": Array [
      1.2345678910111213e+30,
      0,
    ],
  },
]
`;
